fix(avl): add missing checks in avl package

davd-gzl opened the pull request:
Fix: #4440 (already closed)

This follows notJoon comment : `The three issues you pointed out can be resolved with proper bounds checking alone, without requiring separate type changes. In particular, it would be good to add guard for GetByIndex and TraverseByOffset methods.` from https://github.com/gnolang/gno/issues/4440#issuecomment-3035628209

As well I added a check on pager `GetPageWithSize`, as it can trigger a division by zero / undefined behavior.

I verified and there shouldn't be any other needed checks in avl packages.

davd-gzl committed:
fix(avl): add index and size overflow checks in GetByIndex and traversebyoffset

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: davd-gzl/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🟢 User notJoon already reviewed PR 4908 with state APPROVED
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

davd-gzl committed:
fix(avl): add overflow checks for height and size in calcHeightAndSize

davd-gzl committed:
fix(pager): add panic for invalid page number or page size in GetPageWithSize

davd-gzl committed:
fix(avl): remove panic on negative offset

davd-gzl committed:
fix(avl): remove panic on negative limit in TraverseByOffset

davd-gzl committed:
fix(pager): remove check on negative pageNumber, as it should be supported (ref: pager/z_filetest.gno)

Villaquiranm commented on examples/gno.land/p/nt/avl/node.gno:
```diff
@@ -108,12 +108,15 @@ func (node *Node) Get(key string) (index int, value any, exists bool) {
-			return node.key, node.value
-		} else {
+		if index != 0 {
 			panic("GetByIndex asked for invalid index")
```
Same here
nitpick: I think it is better you can have this change reverted to avoid reviewing uneccessary changes

davd-gzl committed:
fix(avl): reverse unnecessary change

davd-gzl committed:
Merge remote-tracking branch 'origin/master' into fix/avl-signed-size

thehowl committed:
Merge branch 'master' into fix/avl-signed-size

thehowl committed:
Merge branch 'master' of github.com:gnolang/gno into fix/avl-signed-size

thehowl committed:
Merge branch 'master' into fix/avl-signed-size

thehowl commented on examples/gno.land/p/nt/avl/v0/node.gno:
The `TraverseByOffset` guard requested in https://github.com/gnolang/gno/issues/4440#issuecomment-3035628209 is still missing, and it's the only one of the three issue points that produces silently **wrong results** rather than a panic:

```go
// 4-leaf tree {"a","b","c","d"}
tree.TraverseByOffset(0, 4, true, true, cb)   // visits [a b c d]
tree.TraverseByOffset(-2, 4, true, true, cb)  // visits [a b] — c, d silently dropped
```

The cause is in `traverseByOffset`: [`delta := first.size - offset`](https://github.com/gnolang/gno/blob/08efe9c3ec9383ff717d9fb44784b10be78f83b0/examples/gno.land/p/nt/avl/v0/node.gno#L449) over-counts when `offset < 0` (it becomes `first.size + |offset|`), so `delta >= limit` returns early before the right subtree is visited. This is reachable from `Tree.IterateByOffset` / `ReverseIterateByOffset`, which the pager itself calls.

Please add a guard in the [`TraverseByOffset` fast path](https://github.com/gnolang/gno/blob/08efe9c3ec9383ff717d9fb44784b10be78f83b0/examples/gno.land/p/nt/avl/v0/node.gno#L395-L398) — either clamp `offset < 0` to 0 (consistent with the lenient `limit <= 0` handling there) or panic (consistent with your new `GetByIndex` guard).

Negative *limit* is already covered by the existing `limit <= 0` check, so nothing needed there.

thehowl commented on examples/gno.land/p/nt/avl/v0/node.gno:
```diff
@@ -108,12 +108,15 @@ func (node *Node) Get(key string) (index int, value any, exists bool) {
 // GetByIndex retrieves the key-value pair of the node at the given index
 // in the subtree rooted at the node.
 func (node *Node) GetByIndex(index int) (key string, value any) {
+	if index < 0 {
+		panic("GetByIndex: negative index not allowed")
+	}
```
This guard looks good. Worth noting it was never memory-unsafe before: a negative index already descended left and panicked at the leaf (`"GetByIndex asked for invalid index"`), so this improves error clarity and exits early — still welcome.

thehowl commented on examples/gno.land/p/nt/avl/v0/node.gno:
```diff
@@ -267,6 +270,13 @@ func (node *Node) rotateLeft() *Node {
+	}
+	if node.size < 0 {
+		panic("size overflow: node size exceeds int max")
+	}
```
I'd drop these two checks: they are unreachable in practice. Height > 127 requires ≥ Fib(130) ≈ 2⁸⁸ nodes, and size overflow requires 2⁶³ leaves — as notJoon noted in the issue, this category is "practically impossible" and "doesn't need to be handled separately".

They also sit on the hottest path of the most-used data structure on the chain (every `Set`/`Remove` calls `calcHeightAndSize` O(log n) times, and each rotation twice more), and can never be exercised by a test.

```suggestion
```

thehowl commented on examples/gno.land/p/nt/avl/v0/pager/pager.gno:
```diff
@@ -53,6 +53,10 @@ func (p *Pager) GetPage(pageNumber int) *Page {
 }
 
 func (p *Pager) GetPageWithSize(pageNumber, pageSize int) *Page {
+	if pageSize <= 0 {
+		panic("GetPageWithSize: invalid page size")
+	}
```
This panic makes the existing graceful branch below ([lines 70–73](https://github.com/gnolang/gno/blob/08efe9c3ec9383ff717d9fb44784b10be78f83b0/examples/gno.land/p/nt/avl/v0/pager/pager.gno#L70-L73)) unreachable dead code:

```go
// pages without content
if pageSize < 1 {
    return page
}
```

I think you can remove that block.

davd-gzl committed:
Merge branch 'master' into fix/avl-signed-size

davd-gzl committed:
fix(avl): clamp negative offset in TraverseByOffset; drop unreachable checks

A negative offset made traverseByOffset compute `delta := first.size -
offset` larger than the left subtree, over-subtracting from limit (or
tripping the `delta >= limit` early return) and silently dropping the
right subtree. Clamp negative offset to 0, consistent with the lenient
limit <= 0 handling, and add a regression test.

Also remove the calcHeightAndSize height/size overflow panics: an AVL
tree whose size fits in int can never reach height int8 max (~88 < 127),
and size overflow needs ~2^63 nodes, so both were unreachable. Drop the
now-dead pageSize < 1 branch in GetPageWithSize, which already panics on
pageSize <= 0 at the top.

davd-gzl commented:
Fix: https://github.com/gnolang/gno/pull/4908/commits/689a4e3b242e4d76866ac0bf3da7b5f5e1134ce5
Thank you for reviewing my older work Morgan! :)

davd-gzl committed:
Merge branch 'master' into fix/avl-signed-size

thehowl committed:
Merge branch 'master' into fix/avl-signed-size
